src/mondemand_backend_annotations_elasticsearch.erl

```diff
@@ -0,0 +1,240 @@
+init ([Config]) ->
+  Number = proplists:get_value (number, Config, erlang:system_info(schedulers)), %Was 16
+
+  { ok,
+    {
+      {one_for_one, 10, 10},
+      [
+        { ?POOL,
+          { mondemand_backend_worker_pool_sup, start_link,
+            [ ?POOL,
+              mondemand_backend_worker,
+              Number,
+              ?MODULE ]
+          },
+          permanent,
+          2000,
+          supervisor,
+          [ ]
+        }
+      ]
+    }
+  }.
+
+%%====================================================================
+%% mondemand_backend_worker callbacks
+%%====================================================================
+create (Config) ->
+  AnnotationLabels = proplists:get_value (annotation_labels, Config),
+  CreateUrl = proplists:get_value (create_url, Config),
+  Timeout = proplists:get_value (timeout, Config),
+  UpdateUrl = proplists:get_value (update_url, Config),
+  {ok, #state { create_url = CreateUrl, update_url = UpdateUrl, timeout = Timeout,
+                annotation_labels = AnnotationLabels
+              }}.
+
+connected (#state { create_url = undefined, update_url = undefined }) -> false;
+connected (_) -> true.
+
+connect (State) ->
+  {ok, State}.
+
+send (State = #state {annotation_labels = AnnotationLabels, create_url = Create_Url,
+                      update_url = Update_Url, timeout = Timeout}, Data) ->
+  Annotation = get_annotation(Data, AnnotationLabels),
+  Url = select_url(Annotation, Create_Url, Update_Url),
+  case catch lhttpc:request
+         (Url, "POST",[], format_annotation(Annotation, AnnotationLabels), Timeout, [{max_connections, 256}])
+  of
+    {ok, {{204,_},_,_}} ->
+      { ok, State };
+    E ->
+      error_logger:error_msg ("elasticsearch responded with ~p",[E]),
+      { error, State }
+  end.
+
+destroy (_) ->
+  ok.
+
+%%======================
+%% Internal functions
+%%======================
+get_annotation ( { _Type, _RecTime, _SenderIp, _SenderPort, _Name, MSG }, Labels ) ->
+  { _MsgName, ID, Timestamp, Description, Title, TagCount, Tags, ContextCount, Context } = MSG,
+  TagStr =
+    case Tags of
+      [] -> <<>>;
+      T when is_list (T), T =/= [] ->
+        iolist_to_binary((mondemand_server_util:join (T,",")))
+    end,
+  [ { timestamp, Timestamp }, { description, Description }, { title, Title },
+    { tags, TagStr }, { id, ID } ].
+
+select_url ( AnnotationList, Create_Url, Update_Url) ->
+  case proplists:get_value(id, AnnotationList) of
+    ID when ID /= <<>> -> lists:flatten(io_lib:format(Update_Url, [ID]));
+    _ -> Create_Url
+  end.
+
+format_annotation ( Annotation, Labels ) ->
+  InterAnnotation = [ { proplists:get_value(Name, Labels, Name), Value }
+                      || { Name, Value } <- Annotation,
+                         is_integer(proplists:get_value( Name, Annotation) ) == false,
+                         proplists:get_value( Name, Annotation) =/= <<>>
+                    ],
```

Comment:
it's to make it so that we don't have to update mondemand-tool whenever the lables used in elasticsearch (or whatever) are changed. We can just change the lables in the config and map whatever comes in from mondemand-tool to the appropriate field labels in the backend.